### PR TITLE
Add IsingXY [unitaryhack]

### LINKED ISSUE #2559: Adding `IsingXY` operation
### **Context**

We would like to define a new operation called IsingXY, as [defined in the PennyLane-Braket plugin](https://amazon-braket-pennylane-plugin-python.readthedocs.io/en/latest/code/api/braket.pennylane_plugin.XY.html).

### **Steps to take**

 **Adding the operation**

- Adding the IsingXY class and its attributes to `pennylane/ops/qubit/parametric_ops.py`. 
- Complete its docstring in a similar fashion as for other operations. Make sure the class is accessible through qml.IsingXY.
- Add its decomposition, matrix, eigenvalues, adjoint and generator.
- Make sure the operation works with parameters from all interfaces (autograd, torch, tensorflow and jax)

**Adding devices support**

- Add support for the operation in the `default.qubit` and `default.mixed` devices

**Testing**

- Add the matrix to `pennylane/tests/gate_data.py`
- Test the gate itself
- Test the gradients with multiple interfaces (autograd, torch, tensorflow and jax) with an **analytical** formula.

### **Notes**

You can refer to the implementation and testings of `IsingXX`, `IsingYY` and `IsingZZ`.

As an example, here is the [implementation](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/qubit/parametric_ops.py#L2246) and [the testing of the decomposition](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/qubit/test_parametric_ops.py#L307) for the `IsingYY` operation. The linked test file contains further test cases for `IsingYY`.

### PR DESCRIPTION
**Pull request by ankit27kh**
**Context:**
Add IsingXY gate
**Description of the Change:**
Adds the gate and related methods and tests.
**Benefits:**
Now supports `IsingXY`
**Possible Drawbacks:**
N/A
**Related GitHub Issues:**
Closes #2559

### CODE REVIEW

**Comment by ankit27kh** [comment]
hey @rmoyard it is ready for review. you can take a look

File: pennylane/ops/qubit/parametric_ops.py [hunk 1]
```diff
@@ -2527,3 +2527,166 @@ def adjoint(self):
...
+    num_wires = 2
+    num_params = 1
+    """int: Number of trainable parameters that the operator depends on."""
+
+    grad_method = "A"
+    parameter_frequencies = [(0.5, 1.0)]
+
+    def generator(self):
+        return qml.Hamiltonian(
+            [0.25, 0.25],
```
Comment (rmoyard) [inline pennylane/ops/qubit/parametric_ops.py:2576]: This issue https://github.com/PennyLaneAI/pennylane/pull/2376 was resolved therefore you do not need the hamiltonian and can write it directly see `IsingXX` for example.

File: pennylane/ops/qubit/parametric_ops.py [hunk 2]
```diff
@@ -2527,3 +2527,166 @@ def adjoint(self):
...
+            phi (float): the phase angle
+            wires (Iterable, Wires): the subsystem the gate acts on
+
+        Returns:
+            list[Operator]: decomposition into lower level operations
+
+        **Example:**
+
+        >>> qml.IsingXY.compute_decomposition(1.23, wires=(0,1))
+        [CY(wires=[0, 1]), RY(0.615, wires=[0]), RX(-0.615, wires=[1]), CY(wires=[0, 1]), Hadamard(wires=[0])]
```
Comment (rmoyard) [inline pennylane/ops/qubit/parametric_ops.py:2605]: This line does not correspond to the decomposition below?

File: pennylane/ops/qubit/parametric_ops.py [hunk 3]
```diff
@@ -2527,3 +2527,166 @@ def adjoint(self):
...
+        Returns:
+            tensor_like: eigenvalues
+
+        **Example**
+
+        >>> qml.IsingXY.compute_eigvals(0.5)
+        array([1.        +0.j        , 1.        +0.j        ,       0.96891242-0.24740396j, 0.96891242+0.24740396j])
+        """
+        if qml.math.get_interface(phi) == "tensorflow":
+            phi = qml.math.cast_like(phi, 1j)
```
Comment (rmoyard) [inline pennylane/ops/qubit/parametric_ops.py:2680]: You need to test `compute_eigvals` for tensorflow.

File: tests/ops/qubit/test_parametric_ops.py [hunk 4]
```diff
@@ -1854,3 +2017,14 @@ def test_pow_matrix(self, op, n):
 def test_control_wires(op, control_wires):
     """Test the ``control_wires`` attribute for parametrized operations."""
     assert op.control_wires == control_wires
+
+
+class TestEigvalCalculation:
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:2022]: I don't think it is the best place for this test, you can simply add one after the isingXY tests, also you have to add multiple interfaces for the the tests (tf, torch and jax), it will solve your coverage problem. Also you can use `qml.math.eigvals` for agnostic function for eigenvalues  and `qml.math.allclose` to compare the results and not `Counter`.

**Review by rmoyard — COMMENTED** [review]
Thank you @ankit27kh, you are on a very good way! It misses some coverage and I added some comments.

File: pennylane/ops/qubit/parametric_ops.py (hunk 2 above)
Comment (ankit27kh) [reply pennylane/ops/qubit/parametric_ops.py]: Ah! this was a typo. fixed

File: pennylane/ops/qubit/parametric_ops.py (hunk 1 above)
Comment (ankit27kh) [reply pennylane/ops/qubit/parametric_ops.py]: done

File: tests/ops/qubit/test_parametric_ops.py (hunk 4 above)
Comment (ankit27kh) [reply tests/ops/qubit/test_parametric_ops.py]: done!

File: pennylane/ops/qubit/parametric_ops.py (hunk 3 above)
Comment (ankit27kh) [reply pennylane/ops/qubit/parametric_ops.py:2848]: done

File: tests/ops/qubit/test_parametric_ops.py (hunk 4 above)
Comment (ankit27kh) [reply tests/ops/qubit/test_parametric_ops.py]: Hey @rmoyard, as the values are complex, there is no direct way to arrange them in ascending or descending order. I instead decided to verify them against the analytic formula for the eigenvalues derived from the matrix directly. This works perfectly.

Comment (rmoyard) [reply tests/ops/qubit/test_parametric_ops.py]: Nice @ankit27kh, I will give a last review round today to be sure that everything is fine 👍

File: tests/ops/qubit/test_parametric_ops.py [hunk 5]
```diff
@@ -526,6 +527,41 @@ def test_isingxx_decomposition(self, tol):
...
+                # RY and Hadamard gate
+                mats.append(np.kron(i.matrix(), np.eye(2)))
+            elif i.wires == Wires([2]):
+                # RX gate
+                mats.append(np.kron(np.eye(2), i.matrix()))
+            else:
+                mats.append(i.matrix())
+
+        decomposed_matrix = np.linalg.multi_dot(mats)
+
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:564]: I think you are missing the assert line here.

File: tests/ops/qubit/test_parametric_ops.py [hunk 6]
```diff
@@ -881,6 +917,87 @@ def get_expected(theta):
...
+        )
+
+        param = np.pi
+        assert np.allclose(qml.IsingXY.compute_matrix(param), get_expected(param), atol=tol, rtol=0)
+        assert np.allclose(
+            qml.IsingXY(param, wires=[0, 1]).matrix(), get_expected(param), atol=tol, rtol=0
+        )
+
+    @pytest.mark.parametrize("phi", np.linspace(-np.pi, np.pi, 10))
+    def test_isingxy_eigvals(self, phi, tol):
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:946]: Pls add a small docstring here 👍

File: tests/ops/qubit/test_parametric_ops.py [hunk 7]
```diff
@@ -881,6 +917,87 @@ def get_expected(theta):
...
+            qml.math.cos(phi / 2) + 1j * qml.math.sin(phi / 2),
+            qml.math.cos(phi / 2) - 1j * qml.math.sin(phi / 2),
+            1,
+            1,
+        ]
+        assert qml.math.allclose(evs, evs_expected)
+
+    @pytest.mark.torch
+    @pytest.mark.parametrize("phi", np.linspace(-np.pi, np.pi, 10))
+    def test_isingxy_eigvals_torch(self, phi, tol):
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:973]: Pls add a small docstring here 👍

File: tests/ops/qubit/test_parametric_ops.py [hunk 8]
```diff
@@ -881,6 +917,87 @@ def get_expected(theta):
...
+            qml.math.cos(phi / 2) + 1j * qml.math.sin(phi / 2),
+            qml.math.cos(phi / 2) - 1j * qml.math.sin(phi / 2),
+            1,
+            1,
+        ]
+        assert qml.math.allclose(evs, evs_expected)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("phi", np.linspace(-np.pi, np.pi, 10))
+    def test_isingxy_eigvals_jax(self, phi, tol):
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:988]: Pls add a small docstring here 👍

File: tests/ops/qubit/test_parametric_ops.py [hunk 9]
```diff
@@ -1654,6 +1837,38 @@ def circuit(phi):
         res = jax.grad(circuit, argnums=0)(phi)
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    @pytest.mark.tf
+    @pytest.mark.parametrize("dev_name,diff_method,phi", configuration)
+    def test_isingxy_tf_grad(self, tol, dev_name, diff_method, phi):
+        """Test the gradient for the gate IsingXY."""
```
Comment (rmoyard) [inline tests/ops/qubit/test_parametric_ops.py:1843]:
```suggestion
        """Test the gradient with Tensorflow for the gate IsingXY."""
```

**Review by rmoyard — COMMENTED** [review]
Hey @ankit27kh, some minor comments to be solved. After that I am happy to approve the PR 💯

File: tests/ops/qubit/test_parametric_ops.py (hunk 5 above)
Comment (ankit27kh) [reply tests/ops/qubit/test_parametric_ops.py:564]: thanks for catching this. fixed now.

**Comment by rmoyard** [comment]
Sorry @ankit27kh we reached a concurrency limit again, can you redo it? Thank you!

**Comment by ankit27kh** [comment]
Hey @rmoyard, it's finished
